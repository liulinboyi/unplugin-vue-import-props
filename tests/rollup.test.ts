/* eslint-disable unicorn/prefer-string-replace-all */
import { resolve } from 'path'
import { rollup } from 'rollup'
import glob from 'fast-glob'
import VueImportProps from '../src/rollup'
import type { Plugin } from 'rollup'

const ToString: Plugin = {
  name: 'to-string',
  transform(code) {
    return `export default \`${code.replace(/`/g, '\\`')}\``
  },
}

async function getCode(file: string, plugins: Plugin[]) {
  const bundle = await rollup({
    input: [file],
    external: ['vue'],
    plugins,
  })
  const output = await bundle.generate({ format: 'esm' })
  return output.output
    .map((file) => {
      if (file.type === 'chunk') {
        return file.code
      } else {
        return file.fileName
      }
    })
    .join('\n')
}

describe('transform', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')
    const files = await glob('tests/fixtures/*.{vue,js,ts}', {
      cwd: root,
      onlyFiles: true,
    })

    for (const file of files.filter((n) => n.endsWith('vue'))) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const unpluginCode = await getCode(filepath, [
          VueImportProps({}),
          ToString,
        ]).catch((err) => err)
        expect(unpluginCode).toMatchSnapshot()
      })
    }
  })
})
