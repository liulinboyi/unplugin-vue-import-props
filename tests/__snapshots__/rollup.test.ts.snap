// Vitest Snapshot v1

exports[`transform > fixtures > tests/fixtures/typescript.vue 1`] = `
"var typescript = \`<script lang=\\"ts\\" setup>
import { } from './index';
defineProps<{name:string;}>()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript1.vue 1`] = `
"var typescript1 = \`<script lang=\\"ts\\" setup>
import { foo } from './index';
foo()
defineProps<{name:string;}>()
foo()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript1 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript2.vue 1`] = `
"var typescript2 = \`<script lang=\\"ts\\" setup>
import { foo } from './index';
foo()
defineProps<{age:number;}>()
foo()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript2 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript3.vue 1`] = `
"var typescript3 = \`<script lang=\\"ts\\" setup>
import { foo } from './index1';
foo()
defineProps<{gender:string;}>()
foo()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript3 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript4.vue 1`] = `
"var typescript4 = \`<script lang=\\"ts\\" setup>
import Test, { foo } from './index1'
foo()
defineProps<{
  name: string
}>()
foo()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript4 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript5.vue 1`] = `
"var typescript5 = \`<script lang=\\"ts\\" setup>

defineProps<{gender:string;}>()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript5 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript6.vue 1`] = `
"var typescript6 = \`<script lang=\\"ts\\" setup>
import { } from './index2';
defineProps<{name:string;}>()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript6 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript7.vue 1`] = `
"var typescript7 = \`<script lang=\\"ts\\" setup>
import { haha } from './index2'
defineProps<haha>()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescript7 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript10.vue 1`] = `
"var typescript10 = \`<template>
  <div></div>
</template>

<script setup lang=\\"ts\\">
import { } from './index5';
const props = defineProps<{name:string;age:number;msg:string;labels:unknown;}>()
console.log(props)
</script>

<style lang=\\"scss\\"></style>
\`;

export { typescript10 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript11.vue 1`] = `
"var typescript11 = \`<template>
  <div></div>
</template>

<script setup lang=\\"ts\\">
import { } from './index5';
const props = withDefaults(defineProps<{name:string;age:number;msg:string;labels:unknown;}>(), {
  msg: () => 'hello',
  labels: () => ['one', 'two'],
})
console.log(props)
</script>

<style lang=\\"scss\\"></style>
\`;

export { typescript11 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript12.vue 1`] = `
"var typescript12 = \`<template>
  <div></div>
</template>

<script setup lang=\\"ts\\">
import { } from './index4';
withDefaults(defineProps<{name:string;age:number;msg:string;labels:unknown;}>(), {
  msg: () => 'hello',
  labels: () => ['one', 'two'],
})
</script>

<style lang=\\"scss\\"></style>
\`;

export { typescript12 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescript13.vue 1`] = `
"var typescript13 = \`<template>
  <div></div>
</template>

<script setup lang=\\"ts\\">
import { } from './index5';
// import { Foo as Test } from \\"./index5\\";
// import Test from './index5'
const props = defineProps<{name:string;age:number;msg:string;labels:unknown;}>()
console.log(props)
</script>

<style lang=\\"scss\\"></style>
\`;

export { typescript13 as default };
"
`;

exports[`transform > fixtures > tests/fixtures/typescriptNormal.vue 1`] = `
"var typescriptNormal = \`<script lang=\\"ts\\" setup>
defineProps<{
  name: string
}>()
console.log('Hello')
</script>

<template>
  <div>
    <h1>Hello World</h1>
  </div>
</template>
\`;

export { typescriptNormal as default };
"
`;
